src/java/com/puppetlabs/jruby_utils/pool/JRubyPool.java

```diff
@@ -134,6 +140,9 @@ public E borrowItem() throws InterruptedException {
             do {
                 if (isPoolLockHeldByAnotherThread(currentThread)) {
                     poolNotLocked.await();
+                } else if (this.poisonPill != null){
+                    // Return the pill immediately if there is one
+                    item = poisonPill;
```

Comment:
> I'm not even sure we need to lock the queue before returning the pill, but it seems good to keep all the logic inside the try/finally block.

I think it would be good to have anything which gets/sets the `poisonPill` member variable be protected by some sort of synchronization mechanism since those actions could happen from different threads around the same time.  Since we already use the `queueLock` for that, seems like it would be okay to continue doing that here.